### DISCUSSION TITLE
Not able to use Jinja template in PostgresOperator sql parameter as filename

### DISCUSSION THREAD

**Discussion by kameshkotwani** [apache/airflow#27092] [Q&A]
### Apache Airflow version

main (development)

### What happened

 I am trying to pass a sql parameter as a Jinja Template, but not able to. I am a bit new to Airflow, here is the code, I am not sure if the Jinja templating works in PostgresOperator or not. If there is any other way we can pass please let me know, again I am not trying to pass the jinja template inside the SQL query or file, I am passing the name of the SQL file using the Jinja template which it is not able to render.

### What you think should happen instead

The Jinja Template should be rendered at `sql` parameter in PostgresOperator.

### How to reproduce

```
write_to_postgres = PostgresOperator(
        task_id ="write_to_postgres",
        postgres_conn_id="mypostgres",
        #TODO investigate, why this is not working
        sql ="queries{{ execution_date.hour }}{{ execution_date.day }}.sql"
    )
``` 
Output of rendered in PostgresOperator
![brave_4cWYODo9ih](https://user-images.githubusercontent.com/36355951/196022184-36dfa60c-81d8-4662-a070-9129ad21b7d6.jpg)

However, the BashOperator is able to render it properly

```bash
  delete_queries_file = BashOperator(
        task_id ="delete_queries_file",
        bash_command="cd /opt/airflow/dags && rm queries{{ execution_date.hour }}{{ execution_date.day }}.sql"
    )
```

Output of the BashOperator with proper rendered of Jinja Template
![brave_lfK8o7p6Av](https://user-images.githubusercontent.com/36355951/196022293-9a5f7c0f-f19e-45b7-91c6-1079e8aaea96.jpg)

UPDATE:
One thing I noticed weird is when I do not put .sql in the `sql` parameter, the template gets rendered properly, which seems weird. Here is the screenshot with Jinja Template rendered without the .sql extension in `sql` parameter, but it works in bash operator as you can seen in the above code which uses .sql at end.

![brave_dMsUZ75GWD](https://user-images.githubusercontent.com/36355951/196025955-4263d95c-3f86-4015-aef0-974b3b55bbab.png)

### Operating System

Windows 10 

### Versions of Apache Airflow Providers

apache-airflow-providers-common-sql==1.2.0
apache-airflow-providers-ftp==3.1.0     
apache-airflow-providers-http==4.0.0    
apache-airflow-providers-imap==3.0.0    
apache-airflow-providers-postgres==5.2.2
apache-airflow-providers-sqlite==3.2.1 

### Deployment

Docker-Compose

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by chenglongyan** [comment]
A string with an ```.sql``` suffix is considered a file rather than a jinjia template.


```python
write_to_postgres = PostgresOperator(
        task_id ="write_to_postgres",
        postgres_conn_id="mypostgres",
        #TODO investigate, why this is not working
        sql ="queries{{ execution_date.hour }}{{ execution_date.day }}.sql"
    )
```
---
Refer to the ```template_ext``` field  of [PostgresOperator](https://airflow.apache.org/docs/apache-airflow-providers-postgres/stable/_api/airflow/providers/postgres/operators/postgres/index.html#airflow.providers.postgres.operators.postgres.PostgresOperator.template_ext) and [templating-with-jinja](https://airflow.apache.org/docs/apache-airflow/stable/tutorial/fundamentals.html#templating-with-jinja)

**Comment by kameshkotwani** [comment]
I am trying to create a filename which is dynamic based on its creation time, so I want to use Jinja Template to specify the filename  dynamically in the `sql` parameter, for example I create a file `queries1215.sql`  (here 12 is hour, 15 is date and it contains multiple queries and saved as a file) using PythonOperator, and the file name changes based on time and date, this works in BashOperator as when deleting the file as seen in the above code, but when same logic is being used in PostgresOperator, template is not rendering. I am not able to achieve the desired result due to that. Is there any other way we can do this where the filename is dynamic??

**Comment by uranusjr** [comment]
Unfortunately this is not possible in the current implementation. Maybe we could support it though, it’s not particularly easy to imagine people are actively naming files like Jinja templates that this feature addition would break compatibility (and we can always add a config for this if we are worried).

In the mean time, I’m changing this to a feature request instead.

**Comment by kameshkotwani** [comment]
Thank you for considering it, meanwhile I was able to do the required task using vanilla python and PythonOperator, I have attached the code, might help someone with same requirement as mine.

```python
def _write_to_postgres(execution_date):
    hour = execution_date.hour
    day = execution_date.day
    filename = f"queries{hour}{day}.sql"
    import psycopg2
    try:
        conn = psycopg2.connect(database="airflow",user='airflow', password='airflow', host='postgres', port= '5432')
    except:
        raise AirflowFailException
    else:
        cursor = conn.cursor()
        with open(f"/opt/airflow/dags/{filename}","r") as f:
            for statement in f.readlines():
                print(statement)
                cursor.execute(statement)
        conn.commit()
        cursor.close()
```

**Comment by eladkal** [comment]
> I am trying to create a filename which is dynamic based on its creation time

What process creates the actual file?

**Comment by kameshkotwani** [comment]
There is a python function which creates the file, here is the code:
```python
def _fetch_pageviews(pagenames,execution_date,**_):
    result = dict.fromkeys(pagenames,0)
    with open(f"/opt/airflow/dags/wikipageviews{execution_date.hour}{execution_date.day}",'r') as f:
        for line in f:
            domain_code,page_title,view_counts,_ = line.split(" ")
            if domain_code == "en" and page_title in pagenames:
                result[page_title] = view_counts
    
    with open(f"/opt/airflow/dags/queries{execution_date.hour}{ execution_date.day }.sql",'w') as f:
        for pagename,pageviewcount in result.items():
            f.write(
                f"INSERT INTO pageview_counts VALUES ('{pagename}' ,{pageviewcount}, '{execution_date}' );\n"
            )

```

**Comment by eladkal** [comment]
I'm not sure how you verify the file will exist on the same worker as the downstream task but lets keep that a side.
In this case you should push the file name in Xcom.

I'm converting this to Discussion as for the moment this falls into utalizing Xcoms.

**Comment by jigyasu-mms** [comment]
have you tried this ?
```
sql = "{% include 'queries{{ execution_date.hour }}{{ execution_date.day }}.sql' %}"
```
